src/agent/src/policy.rs

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) 2023 Microsoft Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use anyhow::{anyhow, Result};
+use reqwest::Client;
+use serde::{Deserialize, Serialize};
+use tokio::time::{sleep, Duration};
+
+#[cfg(feature = "agent-policy-debug")]
+use tokio::io::AsyncWriteExt;
+
+static EMPTY_JSON_INPUT: &str = "{\"input\":{}}";
+
+static OPA_DATA_PATH: &str = "/data";
+static OPA_POLICIES_PATH: &str = "/policies";
+
+/// Convenience macro to obtain the scope logger
+macro_rules! sl {
+    () => {
+        slog_scope::logger()
+    };
+}
+
+/// Example of HTTP response from OPA: {"result":true}
+#[derive(Debug, Serialize, Deserialize)]
+struct AllowResponse {
+    result: bool,
+}
+
+/// Singleton policy object.
+#[derive(Debug)]
+pub struct AgentPolicy {
+    /// When true policy errors are ignored, for debug purposes.
+    allow_failures: bool,
+
+    /// OPA path used to query if an Agent gRPC request should be allowed.
+    /// The request name (e.g., CreateContainerRequest) must be added to
+    /// this path.
+    query_path: String,
+
+    /// OPA path used to add or delete a rego format Policy.
+    policy_path: String,
+
+    /// Client used to connect a single time to the OPA service and reused
+    /// for all the future communication with OPA.
+    opa_client: Client,
+
+    /// "/tmp/policy.txt" log file for policy activity.
+    #[cfg(feature = "agent-policy-debug")]
+    log_file: Option<tokio::fs::File>,
+}
+
+impl AgentPolicy {
+    /// Create AgentPolicy object.
+    pub fn new(opa_uri: &str, coco_policy: &str) -> Result<Self> {
+        Ok(AgentPolicy {
+            allow_failures: false,
+            query_path: opa_uri.to_string() + OPA_DATA_PATH + coco_policy + "/",
+            policy_path: opa_uri.to_string() + OPA_POLICIES_PATH + coco_policy,
+            opa_client: Client::builder().http1_only().build().unwrap(),
```

Comment:
My experience from debugging thousands of crashes at Microsoft shows that it's better to crash early than to propagate a fatal error to the caller and hope they will handle that error correctly. For this kind of error:
- The user wanted to support Policy, since they added Policy support by building using AGENT_POLICY=yes
- The Policy will not work at all, so this is a fatal error for the Agent.

So, I think unwrap is correct here. But, I know that most developers don't have the same debugging experience I have, so don't mind sending these errors to the caller if that's what Kata maintainers prefer.

Reply:
In this case the caller is also part of this PR. It unwraps the output of this function. I don't see any reason to unwrap here when this function returns a result. 

More generally whether an error is fatal and whether an error is passed to the caller are a bit orthogonal. Passing the error up allows us to handle it at the appropriate level and to make the behavior of the callee more closely reflect its signature.

Reply:
I replaced the unwrap calls, but I added an abort call - please see https://github.com/kata-containers/kata-containers/pull/7574/commits/5f1bbe4a79d3d6d07ec25e0253c23b3f6dc758b0